ouroboros-consensus-diffusion/test/consensus-test/Test/Consensus/Genesis/Tests/Uniform.hs

```diff
@@ -0,0 +1,126 @@
+{-# LANGUAGE LambdaCase          #-}
+{-# LANGUAGE NamedFieldPuns      #-}
+{-# LANGUAGE RecordWildCards     #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TupleSections       #-}
+{-# LANGUAGE TypeFamilies        #-}
+{-# LANGUAGE TypeOperators       #-}
+
+module Test.Consensus.Genesis.Tests.Uniform (tests) where
+
+import           Cardano.Slotting.Slot (SlotNo (SlotNo), WithOrigin (..))
+import           Control.Monad.IOSim (runSimOrThrow)
+import           Data.List (intercalate)
+import qualified Data.List.NonEmpty as NonEmpty
+import           Data.Word (Word64)
+import           Ouroboros.Consensus.Block (ChainHash (..), HeaderHash)
+import           Ouroboros.Consensus.Util.Condense (condense)
+import qualified Ouroboros.Network.AnchoredFragment as AF
+import           System.Random.Stateful (runSTGen_)
+import           Test.Consensus.BlockTree (btBranches, btTrunk)
+import           Test.Consensus.Genesis.Setup
+import           Test.Consensus.Genesis.Setup.Classifiers
+import           Test.Consensus.PeerSimulator.Run
+                     (SchedulerConfig (scEnableGdd), noTimeoutsSchedulerConfig)
+import           Test.Consensus.PeerSimulator.StateView
+import           Test.Consensus.PointSchedule
+import           Test.Ouroboros.Consensus.ChainGenerator.Params (Delta (Delta))
+import qualified Test.QuickCheck as QC
+import           Test.QuickCheck
+import           Test.QuickCheck.Extras (unsafeMapSuchThatJust)
+import           Test.QuickCheck.Random (QCGen)
+import           Test.Tasty
+import           Test.Tasty.QuickCheck
+import           Test.Util.Orphans.IOLike ()
+import           Test.Util.QuickCheck (le)
+import           Test.Util.TestBlock (TestHash (TestHash))
+import           Test.Util.TestEnv (adjustQuickCheckTests)
+
+tests :: TestTree
+tests =
+  adjustQuickCheckTests (`div` 10) $
+  testProperty "uniform random schedule" prop_uniformSchedule
+
+genChainsAndSchedule :: PointScheduleConfig -> ScheduleType -> QC.Gen (GenesisTest, PointSchedule)
+genChainsAndSchedule scheduleConfig scheduleType =
+  unsafeMapSuchThatJust do
+    numAdversaries <- QC.choose (1, 4)
+    gt <- genChains numAdversaries
+    seed :: QCGen <- arbitrary
+    pure ((gt,) <$> runSTGen_ seed (\ g -> genSchedule g scheduleConfig scheduleType (gtBlockTree gt)))
+
+simpleHash ::
+  HeaderHash block ~ TestHash =>
+  ChainHash block ->
+  [Word64]
+simpleHash = \case
+  BlockHash (TestHash h) -> reverse (NonEmpty.toList h)
+  -- not matching on @GenesisHash@ because 8.10 can't prove exhaustiveness of
+  -- TestHash with the equality constraint
+  _ -> []
+
+makeProperty ::
+  GenesisTest ->
+  Int ->
+  StateView ->
+  [PeerId] ->
+  Property
+makeProperty genesisTest advCount StateView {svSelectedChain} killed =
+  classify genesisWindowAfterIntersection "Full genesis window after intersection" $
+  classify (isOrigin immutableTipHash) "Immutable tip is Origin" $
+  classify (length killed > 0) "Some peers were killed" $
```

Comment:
Use `label` instead to get the distribution of the nodes that stayed connected (?)

Reply:
e.g.
```diff
--- a/ouroboros-consensus-diffusion/test/consensus-test/Test/Consensus/Genesis/Tests/Uniform.hs
+++ b/ouroboros-consensus-diffusion/test/consensus-test/Test/Consensus/Genesis/Tests/Uniform.hs
@@ -37,6 +37,7 @@ import           Test.Util.Orphans.IOLike ()
 import           Test.Util.QuickCheck (le)
 import           Test.Util.TestBlock (TestHash (TestHash))
 import           Test.Util.TestEnv (adjustQuickCheckTests)
+import           Text.Printf (printf)
 
 tests :: TestTree
 tests =
@@ -73,6 +74,7 @@ makeProperty genesisTest advCount StateView {svSelectedChain} killed =
   classify (length killed > 0) "Some peers were killed" $
   classify (length killed == advCount) "All adversaries were killed" $
   classify (advCount < length (btBranches (gtBlockTree genesisTest))) "Some adversaries performed rollbacks" $
+  label (printf "disconnected %.1f%% of adversaries" (100 * fromIntegral (length killed) / fromIntegral advCount :: Double)) $
   counterexample killedPeers $
   conjoin [
     counterexample ("The immutable tip is not honest: " ++ show immutableTip) $
```
```
ouroboros-consensus
  Genesis tests
    uniform random schedule: OK (0.47s)
      +++ OK, passed 10 tests:
      50% Some peers were killed
      30% Some adversaries performed rollbacks
      20% Full genesis window after intersection
      10% All adversaries were killed
      10% Immutable tip is Origin
      
      50% disconnected 0.0% of adversaries
      20% disconnected 50.0% of adversaries
      10% disconnected 100.0% of adversaries
      10% disconnected 33.3% of adversaries
      10% disconnected 66.7% of adversaries

All 1 tests passed (0.48s)
```

Reply:
✅